core/log/batch_logger.cpp

```diff
@@ -0,0 +1,70 @@
+/*******************************<GINKGO LICENSE>******************************
+Copyright (c) 2017-2023, the Ginkgo authors
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions
+are met:
+
+1. Redistributions of source code must retain the above copyright
+notice, this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in the
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include <ginkgo/core/log/batch_logger.hpp>
+
+
+#include <ginkgo/core/base/array.hpp>
+#include <ginkgo/core/base/batch_multi_vector.hpp>
+#include <ginkgo/core/base/math.hpp>
+
+
+namespace gko {
+namespace batch {
+namespace log {
+
+
+template <typename ValueType>
+void BatchConvergence<ValueType>::on_batch_solver_completed(
+    const array<int>& iteration_count,
+    const array<remove_complex<ValueType>>& residual_norm) const
+{
+    if (this->iteration_count_.get_num_elems() == 0) {
+        this->iteration_count_ = gko::array<int>(
+            iteration_count.get_executor(), iteration_count.get_num_elems());
+    }
+    if (this->residual_norm_.get_num_elems() == 0) {
+        this->residual_norm_ = gko::array<remove_complex<ValueType>>(
+            residual_norm.get_executor(), residual_norm.get_num_elems());
+    }
```

Comment:
If we don't use the host executor, then all of this is unnecessary.

Reply:
It is necessary because in case the input is from a view, then the copy operator deduces that `this->iteration_count_` is non-owning and hence cant be resized.

Reply:
You'r right. But I think our array copy assignment is wrong. For a default constructed array, the copy assignment uses the deleter of other, while the move assignment always uses the default deleter, making it owning (until a bit later in the assignment). I think this is wrong and should be the other way around.

Reply:
No, I believe that is by design. For copying, you need to ensure that the deleter is the same as other. For move, if the deleter is different (like in the case of a view), then you cannot move.

Reply:
I think this is something that should be addressed in another PR.